gnosis/eth/oracles/uniswap_v3.py

```diff
@@ -0,0 +1,182 @@
+import functools
+import logging
+from typing import Optional
+
+from eth_abi.exceptions import DecodingError
+from eth_typing import ChecksumAddress
+from web3.contract import Contract
+from web3.exceptions import BadFunctionCallOutput
+
+from gnosis.util import cached_property
+
+from .. import EthereumClient
+from ..constants import NULL_ADDRESS
+from ..contracts import get_erc20_contract
+from .abis.uniswap_v3 import (
+    uniswap_v3_factory_abi,
+    uniswap_v3_pool_abi,
+    uniswap_v3_router_abi,
+)
+from .oracles import CannotGetPriceFromOracle, PriceOracle
+
+logger = logging.getLogger(__name__)
+
+
+class UniswapV3Oracle(PriceOracle):
+    # https://docs.uniswap.org/protocol/reference/deployments
+    UNISWAP_V3_ROUTER = "0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45"
+
+    # Cache to optimize calculation: https://docs.uniswap.org/sdk/guides/fetching-prices#understanding-sqrtprice
+    PRICE_CONVERSION_CONSTANT = 2**192
+
+    def __init__(
+        self,
+        ethereum_client: EthereumClient,
+        uniswap_v3_router_address: Optional[ChecksumAddress] = None,
+    ):
+        """
+        :param ethereum_client:
+        """
+        self.ethereum_client = ethereum_client
+        self.w3 = ethereum_client.w3
+
+        self.router_address = uniswap_v3_router_address or self.UNISWAP_V3_ROUTER
+        self.factory = self.get_factory()
+
+    @classmethod
+    def is_available(
+        cls,
+        ethereum_client: EthereumClient,
+        uniswap_v3_router_address: Optional[ChecksumAddress] = None,
+    ) -> bool:
+        """
+        :param ethereum_client:
+        :param uniswap_v3_router_address:
+        :return: `True` is Uniswap V3 is available for the EthereumClient provided, `False` otherwise
+        """
+        return ethereum_client.is_contract(
+            uniswap_v3_router_address or cls.UNISWAP_V3_ROUTER
+        )
+
+    @functools.lru_cache(maxsize=5120)
+    def get_decimals(self, token_address: str) -> int:
+        try:
+            return (
+                get_erc20_contract(self.w3, token_address).functions.decimals().call()
+            )
+        except (
+            ValueError,
+            BadFunctionCallOutput,
+            DecodingError,
+        ) as e:
+            error_message = f"Cannot get decimals for token={token_address}"
+            logger.warning(error_message)
+            raise CannotGetPriceFromOracle(error_message) from e
+
+    def get_factory(self) -> Contract:
+        """
+        Factory contract creates the pools for token pairs
+
+        :return: Uniswap V3 Factory Contract
+        """
+        try:
+            factory_address = self.router.functions.factory().call()
```

Comment:
This is calling a function that contents a contract address for factory V3 of UniSwap, my question here is could be in the future this router contents the address of V4 with the next release? I want to be only sure because maybe an update could break some calls.

Reply:
It shouldn't. Our current implementation for Uniswap V2 relies on the router and it only returns factory for the V2. But in any case we will found out with our integration tests 😉